cmd/incusd/main_forkbpf.go

```diff
@@ -0,0 +1,285 @@
+		fprintf(stderr, "Failed to create socket pair: %s", strerror(errno));
+		_exit(EXIT_FAILURE);
+	}
+
+	incus_forkbpf_socket_parent = fds[0];
+	incus_forkbpf_socket_child = fds[1];
+
+	incus_forkbpf_pid = fork();
+	if (incus_forkbpf_pid < 0) {
+		fprintf(stderr, "%s - Failed to create new process\n",
+			strerror(errno));
+		_exit(EXIT_FAILURE);
+	}
+
+	if (incus_forkbpf_pid == 0) {
+		// Attach to the user namespace.
+		snprintf(path, sizeof(path), "/proc/%s/ns/user", pidstr);
+		if (dosetns_file(path) < 0) {
+			_exit(EXIT_FAILURE);
+		}
+
+		// Attach to the user namespace.
+		snprintf(path, sizeof(path), "/proc/%s/ns/mnt", pidstr);
+		if (dosetns_file(path) < 0) {
+			_exit(EXIT_FAILURE);
+		}
+	}
+}
+*/
+import "C"
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"golang.org/x/sys/unix"
+
+	// Used by cgo.
+	_ "github.com/lxc/incus/v6/shared/cgo"
+)
+
+type cmdForkbpf struct {
+	global *cmdGlobal
+}
+
+func (c *cmdForkbpf) command() *cobra.Command {
+	// Main subcommand
+	cmd := &cobra.Command{}
+	cmd.Use = "forkbpf <mount-path> <cmd-types> <map-types> <prog-types> <attach-types>"
+	cmd.Args = cobra.ExactArgs(5)
+	cmd.Short = "Mount bpffs"
+	cmd.Long = `Description:
+  Mount bpffs
+
+  This internal command is used to mount a bpf filesystem into the container for bpf token delegation.
+`
+	cmd.Hidden = true
+	cmd.Run = c.run
+	return cmd
+}
+
+func (c *cmdForkbpf) run(_ *cobra.Command, args []string) {
+	mountPath := args[0]
+	cmdTypes := strings.Split(args[1], ",")
+	mapTypes := strings.Split(args[2], ",")
+	progTypes := strings.Split(args[3], ",")
+	attachTypes := strings.Split(args[4], ",")
+
+	if C.incus_forkbpf_pid == 0 {
+		err := c.runChild(int(C.incus_forkbpf_socket_child), mountPath)
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stdout, "[child]: %v", err)
+			os.Exit(1)
+		}
+	} else {
+		childProc, err := os.FindProcess(int(C.incus_forkbpf_pid))
+		if err != nil {
+			_, _ = fmt.Fprint(os.Stdout, "[parent]: Couldn't find child, assuming it already failed, exiting")
+			os.Exit(1)
+		}
+
+		err = c.runParent(int(C.incus_forkbpf_socket_parent), cmdTypes, mapTypes, progTypes, attachTypes)
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stdout, "[parent]: Encountered error, killing child: %v", err)
+			err2 := childProc.Kill()
+			if err2 != nil {
+				_, _ = fmt.Fprintf(os.Stdout, "[parent]: Failed to kill child: %v", err2)
+			} else {
+				_, err2 = childProc.Wait()
+				if err2 != nil {
+					_, _ = fmt.Fprintf(os.Stdout, "[parent]: Failed to wait for child: %v", err2)
+				}
+			}
+
+			os.Exit(1)
+		}
+
+		procState, err := childProc.Wait()
+		if err != nil {
+			_, _ = fmt.Fprintf(os.Stdout, "[parent]: Failed to wait for child: %v", err)
+		} else {
+			if !procState.Success() {
+				_, _ = fmt.Fprint(os.Stdout, "[parent]: Child process failed")
+				os.Exit(1)
+			}
+
+		}
+	}
+}
+
+func (c *cmdForkbpf) runChild(socket int, mountPath string) error {
+	fsfd, err := unix.Fsopen("bpf", 0)
```

Comment:
I think if this call fails, and the parent has already reached the `Recvmsg()` call, it will hang forever. 
I'm not quite sure how to best work around this.

Maybe the parent should call both waitpid and recvmsg in goroutines and send their results through channels so it can select on them?